### ISSUE TITLE
Pgcat is not working with asyncpg python client and I am getting this error in logs ERROR `pgcat::client] Unexpected code: H`

### ISSUE THREAD

**Issue by chethankumar4046** [postgresml/pgcat#396]
** Pgcat is not working with asyncpg python client, but working psycopg2 python client**
Basically i am going to say pgcat not working with synchronisation with python 
  
**To Reproduce**
1. use this image and deploy either in docker or kubernets `ghcr.io/postgresml/pgcat:latest `
2. And use bellow config to run deployment 
`#
# PgCat config example.
#

#
# General pooler settings
[general]
# What IP to run on, 0.0.0.0 means accessible from everywhere.
host = "0.0.0.0"

# Port to run on, same as PgBouncer used in this example.
port = 6432

# Whether to enable prometheus exporter or not.
enable_prometheus_exporter = true

# Port at which prometheus exporter listens on.
prometheus_exporter_port = 9930

# How long to wait before aborting a server connection (ms).
connect_timeout = 5000

# How much time to give `SELECT 1` health check query to return with a result (ms).
healthcheck_timeout = 1000

# How long to keep connection available for immediate re-use, without running a healthcheck query on it
healthcheck_delay = 30000

# How much time to give clients during shutdown before forcibly killing client connections (ms).
shutdown_timeout = 60000

# For how long to ban a server if it fails a health check (seconds).
ban_time = 60 # seconds

# If we should log client connections
log_client_connections = true

# If we should log client disconnections
log_client_disconnections = true

# Reload config automatically if it changes.
autoreload = false

# TLS
# tls_certificate = "server.cert"
# tls_private_key = "server.key"

# Credentials to access the virtual administrative database (pgbouncer or pgcat)
# Connecting to that database allows running commands like `SHOW POOLS`, `SHOW DATABASES`, etc..
admin_username = "postgres"
admin_password = "password"

# pool
# configs are structured as pool.<pool_name>
# the pool_name is what clients use as database name when connecting
# For the example below a client can connect using "postgres://sharding_user:sharding_user@pgcat_host:pgcat_port/sharded"
[pools.test_db]

# Pool mode (see PgBouncer docs for more).
# session: one server connection per connected client
# transaction: one server connection per client transaction
pool_mode = "session"

# If the client doesn't specify, route traffic to
# this role by default.
#
# any: round-robin between primary and replicas,
# replica: round-robin between replicas only without touching the primary,
# primary: all queries go to the primary unless otherwise specified.
default_role = "any"

# Query parser. If enabled, we'll attempt to parse
# every incoming query to determine if it's a read or a write.
# If it's a read query, we'll direct it to a replica. Otherwise, if it's a write,
# we'll direct it to the primary.
query_parser_enabled = true

# If the query parser is enabled and this setting is enabled, the primary will be part of the pool of databases used for
# load balancing of read queries. Otherwise, the primary will only be used for write
# queries. The primary can always be explicitely selected with our custom protocol.
primary_reads_enabled = true

# So what if you wanted to implement a different hashing function,
# or you've already built one and you want this pooler to use it?
#
# Current options:
#
# pg_bigint_hash: PARTITION BY HASH (Postgres hashing function)
# sha1: A hashing function based on SHA1
#
sharding_function = "pg_bigint_hash"

# Credentials for users that may connect to this cluster
[pools.test_db.users.0]
username = "test_db_test4"
password = "password"

# Maximum number of server connections that can be established for this user
# The maximum number of connection from a single Pgcat process to any database in the cluster
# is the sum of pool_size across all users.
pool_size = 50000

# Maximum query duration. Dangerous, but protects against DBs that died in a non-obvious way.
statement_timeout = 0

# Shard 0
[pools.test_db.shards.0]
# [ host, port, role ]
servers = [
    [ "db-cluster-ha.postgresql.svc.cluster.local", 5432, "Primary" ]
]
# Database name (e.g. "postgres")
database = "test_db"`
3. by running this code you will get issue 
`import psycopg2

# Connect to the PostgreSQL database
conn = psycopg2.connect(
    host="localhost",
    database="temp",
    user="temp_test4",
    password="temp_test4_12345",
    port=6432
)

# Open a cursor to perform database operations
cur = conn.cursor()

# Execute a SQL query
cur.execute("SELECT * FROM test4.alert")

# Fetch the results
rows = cur.fetchall()

# Print the results
for row in rows:
    print(row)

# Close the cursor and the database connection
cur.close()
conn.close()

import asyncio
import asyncpg

async def main():
    # Connect to the PostgreSQL database
    conn = await asyncpg.connect(
        host="localhost",
        database="temp",
        user="temp_test4",
        password="temp_test4_12345",
        port=6432
    )

    # Execute a SQL query
    rows = await conn.fetch("SELECT * FROM test4.alert")

    # Print the results
    for row in rows:
        print(row)

    # Close the database connection
    await conn.close()

asyncio.run(main())

**Comment by levkk** [comment]
Hi,

Thank you for reporting the issue. Could you run `RUN_LOG=debug` and provide us with the output? Also curious what errors you get in Python. I think I have a hunch about what could be wrong (async protocol is slightly different than sync - PgCat actually implements the sync protocol I believe). We may need to listen for both server and client messages at the same time, since async sends multiple queries without waiting for server responses.

**Comment by chethankumar4046** [comment]
@levkk 
As you sugguested i have run deployment with  RUST_LOG=debug here is the log
`[2023-04-10T07:21:56.208307Z DEBUG pgcat::client] Password authentication successful
[2023-04-10T07:21:56.208456Z INFO  pgcat::client] Client 127.0.0.1:57072 connected (plain)
[2023-04-10T07:21:56.498748Z DEBUG pgcat::query_router] Query parser enabled: false
[2023-04-10T07:21:56.498775Z DEBUG pgcat::client] Waiting for connection from pool
[2023-04-10T07:21:56.498780Z DEBUG pgcat::pool] Address { id: 0, host: "db-cluster-ha.postgresql.svc.cluster.local", port: 5432, shard: 0, database: "jeeves", role: Primary, replica_number: 0, address_index: 0, username: "jeeves_test4", pool_name: "jeeves", mirrors: [], stats: AddressStats { total_xact_count: 1, total_query_count: 3, total_received: 666, total_sent: 234, total_xact_time: 0, total_query_time: 2, total_wait_time: 29218, total_errors: 0, avg_query_count: 0, avg_query_time: 0, avg_recv: 41, avg_sent: 14, avg_errors: 0, avg_xact_time: 0, avg_xact_count: 0, avg_wait_time: 1826 } } is ok
[2023-04-10T07:21:56.498805Z DEBUG pgcat::client] Got connection from pool
[2023-04-10T07:21:56.498810Z DEBUG pgcat::client] Client 127.0.0.1:57072 talking to server Address { id: 0, host: "db-cluster-ha.postgresql.svc.cluster.local", port: 5432, shard: 0, database: "jeeves", role: Primary, replica_number: 0, address_index: 0, username: "jeeves_test4", pool_name: "jeeves", mirrors: [], stats: AddressStats { total_xact_count: 1, total_query_count: 3, total_received: 666, total_sent: 234, total_xact_time: 0, total_query_time: 2, total_wait_time: 29221, total_errors: 0, avg_query_count: 0, avg_query_time: 0, avg_recv: 41, avg_sent: 14, avg_errors: 0, avg_xact_time: 0, avg_xact_count: 0, avg_wait_time: 1826 } }
[2023-04-10T07:21:56.498823Z ERROR pgcat::client] Unexpected code: H`


And output of python script is hungs for ever

**Comment by levkk** [comment]
I see. Looks like we just need to implement `H` (Flush) [1]. When receiving that message, we need to write out the client buffer to the server.

[1]: https://www.postgresql.org/docs/current/protocol-message-formats.html

**Comment by levkk** [comment]
Yeah, that PR isn't quite right. I haven't had a chance to take a look at it yet, but if you have some time, a PR implementing support for async is welcome.

**Comment by decarbonise** [comment]
Just wanted to bump this. I'd love to use pgcat but we rely on asyncg so we need this feature before we can do so. Thanks!

**Comment by ayZagen** [comment]
My application with https://github.com/porsager/postgres library is not working also. I tried different pool modes (`session` and `transaction`) with `prepared_statements_cache_size` set to `0` or any number. But still the queries fail with this error.

**Comment by optroodt** [comment]
Adding to this, hoping to give the issue some more attention. 😇 

We are running into the same issue (`pgcat-1  | 2024-08-02T09:15:29.638103Z ERROR ThreadId(07) pgcat::client: Unexpected code: H`) and then hangs.

Our stack:

- FastAPI (async)
- SqlAlchemy
- asyncpg

It works fine using Flask (sync).

We'd love to start using PgCat!

**Comment by pchaseh** [comment]
I have the same problem, missing async support makes this project non-viable for a lot of users so it would be great to have

**Comment by AkritW** [comment]
I also would love to have async support

**Comment by mattaltberg** [comment]
Following

**Comment by steve-marmalade** [comment]
Ooof I just hit this error after investing the time in getting it stood up. How about adding a disclaimer to the `README` for the time being so the prospective users are aware before choosing this package?

**Comment by noamliber-reco** [comment]
> Ooof I just hit this error after investing the time in getting it stood up. How about adding a disclaimer to the `README` for the time being so that prospective users are aware before choosing this package?

Same feeling here. is there any plan to support Async one day?

**Comment by thechane** [comment]
> > Ooof I just hit this error after investing the time in getting it stood up. How about adding a disclaimer to the `README` for the time being so that prospective users are aware before choosing this package?
> 
> Same feeling here. is there any plan to support Async one day?

Same same, agree some warning on async library incompatibility up front would have saved me a bunch of time.

**Comment by ManuelAngel99** [comment]
Same problem here, very disappointed to have spent the time setting it up only to find out that it does not support asyncpg 3hours in
